ext/cl_khr_subgroups.asciidoc

```diff
@@ -2,19 +2,42 @@
+
 [[cl_khr_subgroups]]
 == Subgroups
 
 This section describes the *cl_khr_subgroups* extension.
 
-This extension adds support for implementation-controlled groups of work
-items, known as subgroups.
-Subgroups behave similarly to work groups and have their own sets of
-builtins and synchronization primitives, but subgroups within a work group
-are independent, may make forward progress with respect to each other, and
-may map to optimized hardware structures where that makes sense.
+This extension adds support for implementation-controlled groups of work items, known as subgroups.
+Subgroups behave similarly to work groups and have their own sets of built-ins and synchronization primitives.
+Subgroups within a work group are independent, may make forward progress with respect to each other, and may map to optimized hardware structures where that makes sense.
+
+Subgroups were promoted to a core feature in OpenCL 2.1, however note that:
+
+* The subgroup OpenCL C built-in functions described by this extension must still accessed be as an OpenCL C extension in OpenCL 2.1.
+* Subgroup independent forward progress is an optional device property in OpenCL 2.1, see {CL_DEVICE_SUB_GROUP_INDEPENDENT_FORWARD_PROGRESS}.
+
+[[cl_khr_subgroups-additions-to-chapter-3-of-the-opencl-2.0-specification]]
+=== Additions to Chapter 3 of the OpenCL 2.0 Specification
+
+[[cl_khr_subgroups-additions-to-section-3.2-execution-model]]
+=== Additions to section 3.2 -- Execution Model
+
+Within a work-group work-items may be divided into sub-groups.
+The mapping of work-items to sub-groups is implementation-defined and may be queried at runtime.
+While sub-groups may be used in multi-dimensional work-groups, each subgroup is 1-dimensional and any given work-item may query which sub-group it is a member of.
 
-Subgroups became a core feature in OpenCL 2.1.
+Work items are mapped into subgroups through a combination of compile-time decisions and the parameters of the dispatch.
+The mapping to subgroups is invariant for the duration of a kernel’s execution, across dispatches of a given kernel with the same launch parameters, and from one work-group to another within the dispatch (excluding the trailing edge work-groups in the presence of non-uniform work-group sizes).
+In addition, all sub-groups within a work-group will be the same size, apart from the sub-group with the maximum index which may be smaller if the size of the work-group is not evenly divisible by the size of the sub-group.
+
+Sub-groups execute concurrently within a given work-group and make independent forward progress with respect to each other even in the absence of work-group barrier operations.
+Subgroups are able to internally synchronize using barrier operations without synchronizing with each other.
+
+In the degenerate case, with the extension enabled, a single sub-group must be supported for each work-group.
+In this situation all sub-group scope functions alias their work-group level equivalents.
```

Comment:
Do we want to mention the other extreme case where sub-groups are of size 1?

Reply:
This is a little tricky for `cl_khr_subgroups`: since the extension requires subgroup independent forward progress, the only way the "subgroups are size 1" case works is if work items in a work group (which are equivalently "subgroups of size 1") are guaranteed to make independent forward progress.  It's much easier to support this extreme case in OpenCL 2.1, where subgroup independent forward progress is optional.

I'd lean against describing this extreme case here, in the extension doc, at least not without careful text describing IFP requirements.

Reply:
Hmm, agree, probably more trouble than it's worth.